framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+     * type hierarchy will be used to update the method's return type.
+     */
+    public static void updateMethodReturnTypeInScene(ReturnNode retNode,
+            ClassSymbol classSymbol, MethodTree methodTree,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
+        AnnotatedTypeMirror methodReturnType = atf.getAnnotatedType(methodTree).
+                getReturnType();
+        if (methodReturnType.getExplicitAnnotations().size() > 0) {
+            // We do not infer types if there are explicit annotations.
+            // See https://github.com/typetools/annotation-tools/issues/105
+            return;
+        }
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString() +
+                ".jaif";
+        AScene scene = getScene(jaifPath);
+        AnnotatedTypeMirror returnExprATM = atf.getAnnotatedType(retNode.
+                getTree().getExpression());
+
+        AClass clazz = getJaifClass(classSymbol, scene);
+        if (clazz == null) return; // Anonymous class => Ignore, for now.
+        String methodName = JVMNames.getJVMMethodName(methodTree);
+        AMethod method = getJaifMethod(clazz, methodName, scene);
+        if (method.returnType != null) {
+            updateAnnotationSetInScene(
+                    method.returnType.tlAnnotationsHere, atf,
+                    jaifPath, returnExprATM);
+        }
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene. If the set of
+     * annotations (curAnnos) is empty, then the updated set will be the set of
+     * annotations in newATM.
+     * If curAnnos is not empty, the updated set will be the LUB between
+     * curAnnos and newATM.
+     *
+     * @param curAnnos set of annotations located somewhere in a Scene.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update curAnnos.
+     * @param jaifPath identifies a Scene.
+     * @param newATM the type containing a set of annotations to be
+     * added (or "lubbed") to the Scene.
+     */
+    private static void updateAnnotationSetInScene(Set<Annotation> curAnnos,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM) {
+        if (curAnnos != null && curAnnos.size() > 0) {
+            AnnotatedTypeMirror curAnnosATM = setOfAnnotationsToATM(
+                    curAnnos, atf, false, newATM.getUnderlyingType());
+            if (curAnnosATM != null) {
+                newATM = AnnotatedTypes.leastUpperBound(
+                        atf.getProcessingEnv(), atf, newATM, curAnnosATM);
+            }
+        }
+
+        Set<Annotation> setOfAnnos = atmToSetOfAnnotations(newATM, atf);
+        curAnnos.clear();
+        curAnnos.addAll(setOfAnnos);
+        addModifiedScene(jaifPath);
+    }
+
+    /**
+     * Returns the AField in an AScene, given an AClass and a FieldAccessNode.
```

Comment:
I removed this method after some refactoring.

All .jaif files are read by demand, by they are read only once. Once read, their Scene is stored by this class, and there is no more need to read its .jaif file again. Only write operates can occur to update the .jaif file according to the Scene. 

Since all files are read only once, I see no advantage in reading all of them at the start.